reader_writer_test.go

```diff
@@ -0,0 +1,482 @@
+	var err error
+
+	reader := ion.NewReaderStr(value)
+
+	buf := bytes.Buffer{}
+	writer := ion.NewBinaryWriter(&buf)
+
+	err = writer.BeginStruct()
+	if err != nil {
+		t.Errorf("expected writer.BeginStruct() to execute without errors; %s", err.Error())
+	}
+
+	hw, err := NewHashWriter(writer, newIdentityHasherProvider())
+	if err != nil {
+		t.Errorf("expected NewHashWriter() to successfully create a HashWriter; %s", err.Error())
+	}
+
+	ionHashWriter, ok := hw.(*hashWriter)
+	if !ok {
+		t.Error("expected ionHashWriter to be of type hashWriter")
+	}
+
+	err = ionHashWriter.FieldName("field_name")
+	if err != nil {
+		t.Errorf("expected ionHashWriter.FieldName(\"field_name\") to execute without errors; %s", err.Error())
+	}
+
+	writeToWriterFromReader(t, reader, ionHashWriter)
+
+	actual, err := ionHashWriter.Sum(nil)
+	if err != nil {
+		t.Errorf("expected ionHashWriter.Sum(nil) to execute without errors; %s", err.Error())
+	}
+
+	if !reflect.DeepEqual(actual, expectedBytes) {
+		t.Errorf("expected sum to be %v instead of %v", expectedBytes, actual)
+	}
+
+	err = writer.EndStruct()
+	if err != nil {
+		t.Errorf("expected writer.EndStruct() to execute without errors; %s", err.Error())
+	}
+
+	err = ionHashWriter.Finish()
+	if err != nil {
+		t.Errorf("expected ionHashWriter.Finish() to execute without errors; %s", err.Error())
+	}
+
+	err = writer.Finish()
+	if err != nil {
+		t.Errorf("expected writer.Finish() to execute without errors; %s", err.Error())
+	}
+
+	reader = ion.NewReaderBytes(buf.Bytes())
+
+	if !reader.Next() {
+		err = reader.Err()
+		if err != nil {
+			t.Errorf("expected reader.Next() to execute without errors; %s", err.Error())
+		}
+	}
+
+	err = reader.StepIn()
+	if err != nil {
+		t.Errorf("expected reader.StepIn() to execute without errors; %s", err.Error())
+	}
+
+	hr, err := NewHashReader(reader, newIdentityHasherProvider())
+	if err != nil {
+		t.Errorf("expected NewHashReader() to successfully create a HashReader; %s", err.Error())
+	}
+
+	ionHashReader := hr.(*hashReader)
+	if !ok {
+		t.Fatal("expected hr to be of type hashReader")
+	}
+
+	// List
+	if !ionHashReader.Next() {
+		err = ionHashReader.Err()
+		if err != nil {
+			t.Errorf("expected ionHashReader.Next() to execute without errors; %s", err.Error())
+		}
+	}
+
+	// None
+	if !ionHashReader.Next() {
+		err = ionHashReader.Err()
+		if err != nil {
+			t.Errorf("expected ionHashReader.Next() to execute without errors; %s", err.Error())
+		}
+	}
+
+	actualBytes, err := ionHashReader.Sum(nil)
+	if err != nil {
+		t.Errorf("expected ionHashReader.Sum(nil) to execute without errors; %s", err.Error())
+	}
+
+	if !reflect.DeepEqual(expectedBytes, actualBytes) {
+		t.Errorf("expected sum to be %v instead of %v", expectedBytes, actualBytes)
+	}
+}
+
+// Read all the values in the reader and write them in the writer
+func writeToWriterFromReader(t *testing.T, reader ion.Reader, writer ion.Writer) {
```

Comment:
As much as I would like to take credit for this function, I actually borrowed it from https://github.com/amzn/ion-go/blob/master/ion/integration_test.go#L545. 

Perhaps this function should be defined somewhere in ion-go where it can be exposed to ion-hash-go?

Reply:
Oh I see. Not sure if we should add to the public landscape of the ion-go package though. I was referring to reuse of this method within the ion-hash-go package. This is good then.